source/Plugins.cpp

```diff
@@ -38,14 +64,135 @@ namespace {
 				continue;
 
 			for(const DataNode &child : node)
-				if(child.Size() == 2)
+				if(child.Size() == 3)
 				{
+					lock_guard<mutex> guard(pluginsMutex);
 					auto *plugin = plugins.Get(child.Token(0));
 					plugin->enabled = child.Value(1);
 					plugin->currentState = child.Value(1);
+					plugin->version = child.Token(2);
 				}
 		}
 	}
+
+	bool oldNetworkActivity = false;
+
+	mutex activePluginsMutex;
+	set<string> activePlugins;
+
+	// The maximum size of a plugin in bytes, this will be 1 GB.
+	const int MAX_DOWNLOAD_SIZE = 1000000000;
+
+	// Copy an entry from one archive to the other
+	void CopyData(struct archive *ar, struct archive *aw)
+	{
+		int retVal;
+		const void *buff;
+		size_t size;
+		int64_t offset;
+
+		while(true)
+		{
+			retVal = archive_read_data_block(ar, &buff, &size, &offset);
+			if(retVal == ARCHIVE_EOF)
+				return;
+			if(retVal != ARCHIVE_OK)
```

Comment:
Are you sure there is no error handling needed?

Reply:
what kind of error handling do you mean?

Reply:
I added something to stop the process if an error occurs